src/babylon.engine.ts

```diff
@@ -1578,94 +1589,106 @@
-            this._vrDisplayEnabled = vrDevice;
-            this._vrDisplayEnabled.requestPresent([{ source: this.getRenderingCanvas() }]).then(this._onVRFullScreenTriggered);
+        public enableVR() {
+            if (this._vrDisplay && !this._vrDisplay.isPresenting) {
+                var onResolved = () => {
+                    this.onVRRequestPresentComplete.notifyObservers(true);
+                    this._onVRFullScreenTriggered();
+                };
+                var onRejected = () => {
+                    this.onVRRequestPresentComplete.notifyObservers(false);
+                };
+                
+                this.onVRRequestPresentStart.notifyObservers(this);
+                this._vrDisplay.requestPresent([{ source: this.getRenderingCanvas() }]).then(onResolved).catch(onRejected);
+            }
         }
 
         public disableVR() {
-            if (this._vrDisplayEnabled) {
-                this._vrDisplayEnabled.exitPresent().then(this._onVRFullScreenTriggered);
+            if (this._vrDisplay && this._vrDisplay.isPresenting) {
+                this._vrDisplay.exitPresent().then(this._onVRFullScreenTriggered).catch(this._onVRFullScreenTriggered);
             }
         }
 
         private _onVRFullScreenTriggered = () => {
-            if (this._vrDisplayEnabled && this._vrDisplayEnabled.isPresenting) {
+            if (this._vrDisplay && this._vrDisplay.isPresenting) {
                 //get the old size before we change
                 this._oldSize = new BABYLON.Size(this.getRenderWidth(), this.getRenderHeight());
                 this._oldHardwareScaleFactor = this.getHardwareScalingLevel();
 
                 //get the width and height, change the render size
-                var leftEye = this._vrDisplayEnabled.getEyeParameters('left');
+                var leftEye = this._vrDisplay.getEyeParameters('left');
                 var width, height;
                 this.setHardwareScalingLevel(1);
                 this.setSize(leftEye.renderWidth * 2, leftEye.renderHeight);
             } else {
-                //When the specs are implemented, need to uncomment this.
                 this.setHardwareScalingLevel(this._oldHardwareScaleFactor);
                 this.setSize(this._oldSize.width, this._oldSize.height);
-                this._vrDisplayEnabled = undefined;
             }
         }
 
-        private _getVRDisplays() {
+        private _getVRDisplays(callback) {
             var getWebVRDevices = (devices: Array<any>) => {
-
-                this._vrDisplays = devices.filter(function (device) {
-                    return device instanceof VRDisplay;
-                });
-
-                return this._vrDisplays;
+                this._vrSupported = true;
+                // note that devices may actually be an empty array. This is fine;
+                // we expect this._vrDisplay to be undefined in this case.
+                return this._vrDisplay = devices[0];

```

Comment:
Can't we expect a single host to support more than one device parallel?

Reply:
Right now,  no current browser supports multiple concurrent connected VRDisplays. This will likely change when vendors implement WebVR 2.0.

It is an interesting topic of discussion on whether Babylon should support multiple displays for WebVR 1.1. If it does, I can see two levels of support:
- Giving user choice of which headset to render to, when attempting to Enter VR
- Support  rendering to multiple headsets concurrently

The second choice seems impossible to implement now, as there is no way to validate/test the implementation - so will ignore that.

The first choice is more interesting. I removed the existing support primarily because selecting a VRDisplay based on a name via configuration is risky behavior for cross compatibility reasons - and there is no vendor support for the feature in WebVR 1.1. If a 'choice of headset' feature is implemented, it should probably be done as part of the vrExperienceHelper class via the user interaction when they try to Enter VR. Otherwise, we will have huge compatibility issues with every developer implementing their own selection methods via device names... which will also make it harder to migrate them to a new 2.0 implementation. This being the case, perhaps instead could add a new method to Engine that the vrExperienceHelper could call prior to activating the camera. It could show the user a choice on screen if there is more than 1 display connected:

```
public chooseActiveVRDisplay(chooseFunction: (any[]) => any) {
  if (this._VRDisplay && this._VRDisplay.isPresenting) {
    // Throw error?
    // Also throw error if we are waiting for a requestPresent call to resolve?
  }
  // Store a list of all connected displays on Engine. 
  var callback = (newDisplay) => {
    if (!newDisplay) { return; }
    if (!this._vrDisplays.contains(newDisplay)) { return; }
 
    if (newDisplay !== this._VRDisplay) {
      this._VRDisplay = newDisplay; 

      // Notify watchers that the display changed
      var eventArgs = {
        vrDisplay: this._vrDisplay,
        vrSupported: this._vrSupported
      };
      this.onVRDisplayChangedObservable.notifyObservers(eventArgs);
    }
  }
  var newDisplay = chooseFunction(this._vrDisplays, callback);
}
```

Of course - this doesn't solve the issue of automatically launching VR on page load (on browsers that support this). For this case, it might be best to stick with the convention adopted by other implementers and simply enter VR on device 0.